lib/prelude.dx

```diff
@@ -537,6 +545,30 @@ def sq  [Mul a] (x:a) : a = x * x
 def abs [Add a, Ord a] (x:a) : a = select (x > zero) x (zero - x)
 def mod (x:Int) (y:Int) : Int = rem (y + rem x y) y
 
+instance [Floating a] Floating (n=>a)
```

Comment:
Ha, for some reason I never thought to do this. It's a little touch of numpy-style rank polymorphism. I can't see a reason not to allow it.

Reply:
Woah @ the rank polymorphism relevation. Reminds me of typeclass instances for `(->)` in Haskell, e.g. ["How QuickCheck generates random functions"](https://kseo.github.io/posts/2016-12-14-how-quick-check-generate-random-functions.html):

```haskell
class CoArbitrary a where
  coarbitrary :: a -> Gen b -> Gen b

-- | Modifies a generator using an integer seed.
variant :: Int -> Gen a -> Gen a

instance CoArbitrary Bool where
  coarbitrary False = variant 0
  coarbitrary True = variant 1

instance CoArbitrary a => CoArbitrary (Maybe a) where
  coarbitrary Nothing  = variant 0
  coarbitrary (Just x) = variant 1 . coarbitrary x

-- Finally, an answer to the question: how does QuickCheck generate random functions?
instance (Coarbitrary a, Arbitrary b) => Arbitrary (a -> b) where
  arbitrary = promote (\a -> coarbitrary a arbitrary)
```